### DISCUSSION TITLE
How to get TopologyManager using DI or builder in ASP.NET Core App

### DISCUSSION THREAD

**Discussion by pgarbo** [Havret/dotnet-activemq-artemis-client#398] [Q&A] [answered]
Hi,

I am using ArtemisNetClient.Extensions.DependencyInjection in ASP.NET Core application and I need to call ITopologyManager.DeclareQueueAsync before sending any message (to ensure last-value queue is created).  How can I get access to the topology manager using DI or IActiveMqBuilder and where is the best way to call it? 

thanks,
Pawel

**Comment by Havret** [comment]
Hi @pgarbo,

To keep things simple I never exposed `ITopologyManager` as part of `Extensions.DependencyInjection`. You could use this interface implicitly by registering consumer or producer with custom configuration:

https://github.com/Havret/dotnet-activemq-artemis-client/blob/49f7056d4b3b64e27c8a9e905e1dd9d1e62bef34/src/ArtemisNetClient.Extensions.DependencyInjection/ActiveMqArtemisClientDependencyInjectionExtensions.cs#L222

Then you would need to enable address and/or queue declaration:

```csharp
activeMqBuilder.AddConsumer(address, RoutingType.Multicast, queue, new QueueOptions
    {
        LastValueKey = "my-key",
        NonDestructive = true,
    }, MessageHandler)
    .EnableAddressDeclaration()
    .EnableQueueDeclaration();
```

In most of the use cases I've seen so far you want to configure address as part of your producer application, and you want to configure queue as part of your consumer application. It your case it clearly doesn't seem to be the case.  I guess we could add the option to register custom `ConfigureTopology` delegate that will give you access to `ITopologyManager` and will be executed as part of the app startup, right before producers and consumers are brought to life. We have [sth similar already available for connection configuration](https://github.com/Havret/dotnet-activemq-artemis-client/discussions/305#discussioncomment-1683866).

Please let me know if sth along these lines will work for you.

```csharp
activeMqBuilder.ConfigureTopology(async (_, topologyManager) =>
{
    await topologyManager.CreateQueueAsync(new QueueConfiguration
    {
        Address = addressName,
        Name = queueName,
        AutoCreateAddress = true
    });
});
```

Thanks,
Havret

**Reply by pgarbo** [reply]
Hi,

Yes, ConfigureTopology delegate would be ideal.

Thanks,
Pawel

**Comment by Havret** [comment, marked as answer]
Please check [2.11.0](https://www.nuget.org/packages/ArtemisNetClient/2.11.0).
